src/charm_state.py

```diff
@@ -0,0 +1,141 @@
+
+    Attrs:
+        address: The Jenkins server address to register to.
+        secret: The secret used to register agent.
+    """
+
+    address: str
+    secret: str
+
+
+class CharmStateBaseError(Exception):
+    """Represents error with charm state."""
+
+
+class InvalidStateError(CharmStateBaseError):
+    """Exception raised when state configuration is invalid."""
+
+    def __init__(self, msg: str = ""):
+        """Initialize a new instance of the InvalidStateError exception.
+
+        Args:
+            msg: Explanation of the error.
+        """
+        self.msg = msg
+
+
+def _get_jenkins_unit(
+    all_units: typing.Set[ops.Unit], current_app_name: str
+) -> typing.Optional[ops.Unit]:
+    """Get the Jenkins charm unit in a relation.
+
+    Args:
+        all_units: All units in a relation.
+        current_app_name: The Jenkins-agent application name.
+
+    Returns:
+        The Jenkins server application unit in the relation if found. None otherwise.
+    """
+    for unit in all_units:
+        # if the unit's application name is the same, this is peer unit. Otherwise, it is the
+        # Jenkins server unit.
+        if unit.app.name == current_app_name:
+            continue
+        return unit
+    return None
+
+
+def _get_credentials_from_agent_relation(
+    server_unit_databag: ops.RelationDataContent, unit_name: str
+) -> typing.Optional[Credentials]:
+    """Import server metadata from databag in agent relation.
+
+    Args:
+        server_unit_databag: The relation databag content from agent relation.
+        unit_name: The agent unit name.
+
+    Returns:
+        Metadata if complete values(url, secret) are set. None otherwise.
+    """
+    address = server_unit_databag.get("url")
+    secret = server_unit_databag.get(f"{unit_name}_secret")
+    if not address or not secret:
+        return None
+    return Credentials(address=address, secret=secret)
+
+
+@dataclass
+class State:
+    """The Jenkins agent state.
+
+    Attrs:
+        agent_meta: The Jenkins agent metadata to register on Jenkins server.
+        agent_relation_credentials: The full set of credentials from the agent relation. None if
+            partial data is set or the credentials do not belong to current agent.
+        jenkins_agent_service_name: The Jenkins agent workload container name.
+    """
+
+    agent_meta: metadata.Agent
+    agent_relation_credentials: typing.Optional[Credentials]
+    jenkins_agent_service_name: str = "jenkins-agent"
+
+    @classmethod
+    def from_charm(cls, charm: ops.CharmBase) -> "State":
+        """Initialize the state from charm.
+
+        Args:
+            charm: The root Jenkins agent charm.
+
+        Raises:
+            InvalidStateError: if invalid state values were encountered.
+
+        Returns:
+            Current state of Jenkins agent.
+        """
+        try:
+            agent_meta = metadata.Agent(
+                num_executors=os.cpu_count() or 0,
```

Comment:
Isn't this also an error? How useful is a charm with no executors?

Reply:
Yes, this was copied over from the k8s agent and actually has no effect since the pydantic validator used requires this to be >= 1. I've removed it

Reply:
What was removed? It seems this line of code is still here?

Reply:
Oops, the code was not pushed. it's removed now